File: python_version/bot.py
"""
Bot pour la plateforme awale-arena
Format compatible:
- Entrée: mouvements au format "NX" ou "NTX" (N=numéro trou, X=R/B/T, T=transparent)
- Sortie: coup au même format
- Utilise MinMax AI de ai_algorithms.py
"""

import sys
import random
from typing import List, Optional, Tuple
from game_rules import GameState, Color
from game_engine import GameEngine, MoveGenerator
from ai_algorithms import MinMaxBot


def main():
    """
    Fonction principale pour interfacer avec awale-arena
    Lire depuis stdin, traiter les mouvements, envoyer les réponses
    """
    # Récupérer le numéro du joueur (1 ou 2)
    my_player = int(sys.argv[1])  # 1 ou 2

    # Initialiser le jeu et le bot avec MinMax AI
    state = GameState()
    engine = GameEngine(state)
    bot = MinMaxBot(depth=4)  # MinMax avec profondeur 4

    # Boucle principale
    for line in sys.stdin:
        move_str = line.strip()

        # Traiter les signaux spéciaux
        if move_str == "END":
            break

        # Pour Player 2, attendre le premier coup de Player 1
        if move_str == "START" and my_player == 2:
            continue  # Attendre le prochain message

        if move_str != "START":
            # Appliquer le coup de l'adversaire
            hole, color, trans_as = parse_move(move_str)
            if hole and color:
                engine.play_move(hole, color, trans_as)

        # S'assurer que le joueur courant est correct
        state.current_player = my_player
        
        # Obtenir le meilleur coup avec MinMax
        best_move = bot.get_move(state, my_player)

        if best_move:
            hole, color, trans_as = best_move
            my_move = format_move(hole, color, trans_as)
            
            # Appliquer notre coup
            engine.play_move(hole, color, trans_as)
        else:
            # Pas de coup disponible (ne devrait pas arriver)
            my_move = "NOMOVE"

        # Envoyer le coup à la plateforme
        sys.stdout.write(my_move + "\n")
        sys.stdout.flush()


def parse_move(move_str: str) -> Tuple[Optional[int], Optional[Color], Optional[Color]]:
    """
    Parse un mouvement au format "NX" ou "NTX"
    Retourne (hole, color, transparent_as)
    """
    try:
        move_str = move_str.strip().upper()
        
        if 'T' in move_str[1:-1]:  # Transparent au milieu (ex: "3TR", "3TB")
            hole = int(move_str[:-2])
            trans_as_str = move_str[-1]
            color = Color.TRANSPARENT
            trans_as = Color.RED if trans_as_str == 'R' else Color.BLUE
            return (hole, color, trans_as)
        else:  # Normal (ex: "3R", "4B")
            hole = int(move_str[:-1])
            color_str = move_str[-1]
            color = Color.RED if color_str == 'R' else Color.BLUE
            return (hole, color, None)
    except:
        return (None, None, None)


def format_move(hole: int, color: Color, trans_as: Optional[Color]) -> str:
    """
    Formate un mouvement pour l'envoi
    """
    if color == Color.TRANSPARENT:
        return f"{hole}T{trans_as.value}"
    else:
        return f"{hole}{color.value}"


if __name__ == "__main__":
    main()

File: c_version/bot.cpp
/**
 * Bot pour la plateforme awale-arena - Version C++
 * Format compatible:
 * - Entrée: mouvements au format "NX" ou "NTX" (N=numéro trou, X=R/B/T, T=transparent)
 * - Sortie: coup au même format
 * - Utilise MinMax AI de ai_algorithms.h
 */

#include "game_rules.h"
#include "game_engine.h"
#include "ai_algorithms.h"
#include <iostream>
#include <string>
#include <sstream>
#include <tuple>
#include <optional>

// Structure pour représenter un coup parsé
struct ParsedMove
{
    int hole;
    Color color;
    std::optional<Color> trans_as;
    bool valid;
};

// Parse un mouvement au format "NX" ou "NTX"
ParsedMove parse_move(const std::string &move_str)
{
    ParsedMove result = {0, Color::RED, std::nullopt, false};

    try
    {
        std::string upper_move = move_str;
        for (auto &c : upper_move)
            c = std::toupper(c);

        // Chercher 'T' au milieu (transparent)
        if (upper_move.length() >= 3 && upper_move[upper_move.length() - 2] == 'T')
        {
            // Format "NTX" (ex: "3TR", "3TB")
            result.hole = std::stoi(upper_move.substr(0, upper_move.length() - 2));
            result.color = Color::TRANSPARENT;

            char trans_as_char = upper_move[upper_move.length() - 1];
            result.trans_as = (trans_as_char == 'R') ? Color::RED : Color::BLUE;
            result.valid = true;
        }
        else if (upper_move.length() >= 2)
        {
            // Format "NX" (ex: "3R", "4B")
            result.hole = std::stoi(upper_move.substr(0, upper_move.length() - 1));

            char color_char = upper_move[upper_move.length() - 1];
            result.color = (color_char == 'R') ? Color::RED : Color::BLUE;
            result.valid = true;
        }
    }
    catch (...)
    {
        result.valid = false;
    }

    return result;
}

// Formate un mouvement pour l'envoi
std::string format_move(int hole, Color color, std::optional<Color> trans_as)
{
    std::stringstream ss;
    ss << hole;

    if (color == Color::TRANSPARENT && trans_as.has_value())
    {
        ss << "T" << (trans_as.value() == Color::RED ? "R" : "B");
    }
    else
    {
        ss << (color == Color::RED ? "R" : "B");
    }

    return ss.str();
}

int main(int argc, char *argv[])
{
    if (argc < 2)
    {
        std::cerr << "Usage: bot <player_number>" << std::endl;
        return 1;
    }

    // Récupérer le numéro du joueur (1 ou 2)
    int my_player = std::stoi(argv[1]);

    // Initialiser le jeu et le bot avec MinMax AI
    GameState state;
    GameEngine engine(&state); // Pass pointer to GameState
    MinMaxBot bot(4);          // MinMax avec profondeur 4

    std::string line;

    // Boucle principale
    while (std::getline(std::cin, line))
    {
        // Enlever les espaces blancs
        line.erase(0, line.find_first_not_of(" \t\r\n"));
        line.erase(line.find_last_not_of(" \t\r\n") + 1);

        // Traiter les signaux spéciaux
        if (line == "END")
        {
            break;
        }

        // Pour Player 2, on doit d'abord recevoir le coup de Player 1
        if (line == "START" && my_player == 2)
        {
            continue; // Attendre le prochain message (le coup de Player 1)
        }

        if (line != "START")
        {
            // Appliquer le coup de l'adversaire
            ParsedMove parsed = parse_move(line);
            if (parsed.valid)
            {
                if (parsed.color == Color::TRANSPARENT && parsed.trans_as.has_value())
                {
                    engine.playMove(parsed.hole, parsed.color, parsed.trans_as.value(), true);
                }
                else
                {
                    engine.playMove(parsed.hole, parsed.color, Color::RED, false);
                }
            }
        }

        // S'assurer que le joueur courant est correct
        state.current_player = my_player;

        // Obtenir le meilleur coup avec MinMax
        Move best_move = bot.getMove(state, my_player);

        std::string my_move;

        if (best_move.valid)
        {
            std::optional<Color> trans_opt = best_move.use_transparent ? std::optional<Color>(best_move.transparent_as) : std::nullopt;
            my_move = format_move(best_move.hole, best_move.color, trans_opt);

            // Appliquer notre coup
            engine.playMove(best_move.hole, best_move.color,
                            best_move.transparent_as, best_move.use_transparent);
        }
        else
        {
            // Pas de coup disponible (ne devrait pas arriver)
            my_move = "NOMOVE";
        }

        // Envoyer le coup à la plateforme
        std::cout << my_move << std::endl;
        std::cout.flush();
    }

    return 0;
}
